infrastructure/smart-contracts/contracts/upgrades/node-license/NodeLicense8.sol

```diff
@@ -0,0 +1,748 @@
+        require(_xaiAddress != address(0), "Invalid xai address");
+        require(_esXaiAddress != address(0), "Invalid esXai address");
+        require(ethPriceFeedAddress != address(0), "Invalid ethPriceFeed address");
+        require(xaiPriceFeedAddress != address(0), "Invalid xaiPriceFeed address");
+        ethPriceFeed = IAggregatorV3Interface(ethPriceFeedAddress);
+        xaiPriceFeed = IAggregatorV3Interface(xaiPriceFeedAddress);
+        xaiAddress = _xaiAddress;
+        esXaiAddress = _esXaiAddress;
+
+        // Set max supply to 0 to prevent minting until airdrop is complete and pricing tiers are set
+        maxSupply = 0;
+
+        // Delete Existing Pricing Tiers
+        // This is required to ensure when the new pricing tiers are set, the maxSupply is updated correctly
+        delete pricingTiers;
+
+        // Grant the airdrop admin role to the airdrop admin address
+        _grantRole(AIRDROP_ADMIN_ROLE, airdropAdmin);
+    }
+
+    /**
+     * @notice Creates a new promo code.
+     * @param _promoCode The promo code.
+     * @param _recipient The recipient address.
+     */
+    function createPromoCode(string calldata _promoCode, address _recipient) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_recipient != address(0), "Recipient address cannot be zero");
+        require(_promoCodes[_promoCode].recipient == address(0), "Promo code already exists");
+        require(bytes(_promoCode).length > 0, "Promo code cannot be empty");
+        _promoCodes[_promoCode] = PromoCode(_recipient, true, 0);
+        emit PromoCodeCreated(_promoCode, _recipient);
+    }
+
+    /**
+     * @notice Disables a promo code.
+     * @param _promoCode The promo code to disable.
+     */
+    function removePromoCode(string calldata _promoCode) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_promoCodes[_promoCode].recipient != address(0), "Promo code does not exist");
+        _promoCodes[_promoCode].active = false; // 'active' is set to false
+        emit PromoCodeRemoved(_promoCode);
+    }
+
+    /**
+     * @notice Returns the promo code details.
+     * @param _promoCode The promo code to get.
+     * @return The promo code details.
+     */
+    function getPromoCode(string calldata _promoCode) external view returns (PromoCode memory) {
+        return _promoCodes[_promoCode];
+    }
+
+    /**
+     * @notice Returns the length of the pricing tiers array.
+     * @return The length of the pricing tiers array.
+     */
+    function getPricingTiersLength() external view returns (uint256) {
+        return pricingTiers.length;
+    }
+
+    /**
+     * @notice Mints new NodeLicense tokens.
+     * @param _amount The amount of tokens to mint.
+     * @param _promoCode The promo code.
+     */
+    function mint(uint256 _amount, string calldata _promoCode) public payable {
+        // Validate the mint data
+        _validateMint(_amount, _promoCode);
```

Comment:
This is checking if the recipient of the promocode is not msg.sender, but if the promocode is a wallet address and does not exist yet I could be my own recipient right ?

Reply:
This has been fixed by moving the address checks into the calculateReferralRewards function.